### PR TITLE
scylla: lwt optimization to use primary replicas constistently (with CQL protocol extension negotiation)

### PR DESCRIPTION
**Pull request by ManManson**
The driver now ensures that replicas are never shuffled for LWT type queries.

This PR is based on https://github.com/scylladb/gocql/pull/41 from @dahankzter.
The core logic of the previous patch stays the same (e.g. host-selection policy-related stuff and tests).
What differs is that now lwt optimization feature is being negotiated by the new mechanism, which is being incorporated into scylla.
The value for bitmask to search for in prepared statement metadata is also being adjusted when negotiation procedure is complete.

The brief summary of changes from the original PR:
 * Since we would like to employ a negotiation procedure for CQL protocol extensions between
   client driver and scylla server I've added `cqlProtocolExtension` interface and one impl extension,
   also done some wiring all the way along. This negotiation is done by agreeing on negotiated
   feature set passed through SUPPORTED/STARTUP messages betwen server and client.
 * The value for lwt bitmask to test against in `preparedMetadata` is now determined at runtime and
   passed from the server side (this would easily allow us to change the flag value if, for example,
   C* devs decide to implement the same feature in the future, and allows to eliminate hard-coding
   in the drivers' code).
 * `lwt` flag is only updated and used in case we've negotiated lwt feature with the server (and
   similarly: the server only modifies preparedMetadata flags only if the driver responded that
   it understands this extension).

Fixes: #40

### CODE REVIEW

File: frame.go [hunk 1]
```diff
@@ -402,9 +402,11 @@ type framer struct {
 	wbuf []byte
 
 	customPayload map[string][]byte
+
+	flagLWT int
 }
 
-func newFramer(r io.Reader, w io.Writer, compressor Compressor, version byte) *framer {
+func newFramer(r io.Reader, w io.Writer, compressor Compressor, version byte, cqlProtoExts []cqlProtocolExtension) *framer {
```
Comment (mmatczuk) [inline frame.go:409]: I do not like the fact we change framer constructor signature. 
Why we want to conditionally change framer?
If we really do want to do that there are better ways of implementing that.

**Comment by mmatczuk** [comment]
@ManManson what it the rationale behind it? Negotiating flag bitmaps feels like overengineering.

**Comment by ManManson** [comment]
> @ManManson what it the rationale behind it? Negotiating flag bitmaps 
feels like overengineering.

Actually, it's more about negotiating whether to use lwt feature or not, 
rather than negotiating the flag value (this one comes just as a cheap 
and handy bonus). I will expand on the reasons behind this whole thing 
below.

The main reason I have decided to implement protocol extensions 
negotiation procedure in the first place is that while implementing LWT 
metadata flag on the Scylla side, after a few rounds of review (you can 
find it all on the `scylladb-dev` mailing list under the topic mentioned 
in the PR description, but I am going to simplify the task and give a 
brief summary in this response) it was fairly pointed out that:

 1. We can't just unconditionally add a flag and modify flags in the 
    statement metadata since we cannot guarantee that every driver will
    respond correctly to an unknown flag. Though the CQL spec states that
    driver should ignore unknown flags, we can't know for sure 
    that every driver follows the spec accurately (and there are many 
    drivers that we know about and probably even more that we are not 
    even aware of).
 2. The same applies to virtually every possible extension of CQL 
    protocol from our side. There are just two of them for now (LWT 
    metadata flag and shard-awareness) but as we are going to evolve 
    further it's not impossible that we introduce some other 
    optimizations from our end.
 4. In order to safely add new extensions we need to negotiate 
    used extensions between client and server on a per-connection basis 
    (since we can possibly have a cluster with slightly different 
    versions of scylla, for example).
 3. And so it's long desired for us to have a well defined negotiation 
    procedure based on the extensibility features of the CQL protocol 
    itself. That was chosen to be the time to introduce such procedure 
    and LWT feature is the first to use it (it's also planned to move 
    shard-awareness to the same negotiation in some future).
 4. CQL protocol allows to be extended by the means of negotiating 
    extensions via extending `SUPPORTED/STARTUP` messages payload to 
    include some new negotiable capabilities.
 5. SUPPORTED message payload consists of a string multimap, e.g. the 
    same as Map<String, List<String>> (in terms of Java), so that each 
    extension can also supply arbitrary number of arguments, that allow 
    for additional configuration for the extension).
 6. In our case we have decided to un-hardcode the exact value for 
    bitmask to look upon in the driver and pass it as a parameter to 
    the extension for the following reasons:
     * There is an associated feature-request in C*: 
       https://issues.apache.org/jira/browse/CASSANDRA-15746.
       In case C* team decides to add such flag in the future, we would
       want to synchronize the value for the flag with them
       (request from Avi).
     * If we have a configurable value, that the driver does not know 
       about, then we don't have to hardcode the value into each of our
       drivers (gocql, java, c++ is coming, probably rust in the 
       nearest future, etc) and adjust accordingly in order to follow 
       C*.
     * This also means that we don't even need to make a new release 
       for any driver and then instruct our users to check/update 
       scylla or driver versions to use, because the driver will start 
       to use the new value automatically as soon as scylla learns of 
       the new value.

File: frame.go (hunk 1 above)
Comment (ManManson) [reply frame.go:409]: > I do not like the fact we change framer constructor signature.
> Why we want to conditionally change framer?
> If we really do want to do that there are better ways of implementing that.

Can you please expand on this one? Why changing the signature is bad?
I can't yet think of another way to customize serialization/deserialization
procedure based on which CQL protocol extensions we use.

It would be very nice if you could point me to some other way.

Comment (dahankzter) [reply frame.go:409]: One way would be to just move the check out of the constructor and apply it after the initial initialization perhaps in another function. Afaict all the old tests use `nil` for the last argument. If we have tests for lwt at the framer level they can apply the modification accordingly.

I am unsure why changing the framer constructor is bad in itself other than changing upstream code.

File: frame.go [hunk 2]
```diff
@@ -438,6 +440,10 @@ func newFramer(r io.Reader, w io.Writer, compressor Compressor, version byte) *f
 	f.header = nil
 	f.traceID = nil
 
+	if lwtExt := findCQLProtoExtByName(cqlProtoExts, lwtAddMetadataMarkKey); lwtExt != nil {
```
Comment (dahankzter) [inline frame.go:443]: Perhaps a new function that decorates the framer accordingly after initial construction?

File: scylla.go [hunk 3]
```diff
@@ -4,17 +4,96 @@ import (
 	"fmt"
 	"math"
 	"strconv"
+	"strings"
 	"sync/atomic"
 )
 
 // scyllaSupported represents Scylla connection options as sent in SUPPORTED
 // frame.
+// FIXME: Should also follow `cqlProtocolExtension` interface.
```
Comment (dahankzter) [inline scylla.go:13]: What would it bring to have the `scyllaSupported` struct also implementing the `cqlProtocolExtension` interface? When would this be used?

File: frame.go (hunk 2 above)
Comment (ManManson) [reply frame.go:443]: Yes, will do, thanks.

File: scylla.go (hunk 3 above)
Comment (ManManson) [reply scylla.go:13]: It should not be hard to do, but I don't want to include it there since it's completely unrelated to the main cause of this patch (and it's not even anywhere on my priority list at the moment). And it for sure will trigger a lengthy discussion, which will cause unneeded delay for this series to be merged.

File: conn.go [hunk 4]
```diff
@@ -636,7 +645,7 @@ func (c *Conn) recv(ctx context.Context) error {
 		return fmt.Errorf("gocql: frame header stream is beyond call expected bounds: %d", head.stream)
 	} else if head.stream == -1 {
 		// TODO: handle cassandra event frames, we shouldnt get any currently
-		framer := newFramer(c, c, c.compressor, c.version)
+		framer := newFramer(c, c, c.compressor, c.version, c.cqlProtoExts)
```
Comment (mmatczuk) [inline conn.go:648]: Please do not change Framer constructor, it makes this patch unnecessarily big. 
Add a new constructor `newFramerWithExts` or use a plain setter that needs to be called before one starts using the framer.

File: frame.go (hunk 1 above)
Comment (mmatczuk) [reply frame.go:409]: It makes this patch unnecessarily big and you need to change all existing usages for no good reason.
In other words this increases chances of conflicts with upstream changes.

I suggest to add a new constructor newFramerWithExts or use a plain setter that needs to be called before one starts using the framer.

File: scylla.go [hunk 5]
```diff
@@ -4,17 +4,96 @@ import (
...
+// "LWT prepared statements metadata mark" CQL Protocol Extension.
+// This extension, if enabled (properly negotiated), allows Scylla server
+// to set a special bit in prepared statements metadata, which would indicate
+// whether the statement at hand is LWT statement or not.
+//
+// This is further used to consistently choose primary replicas in a predefined
+// order for these queries, which can reduce contention over hot keys and thus
+// increase LWT performance.
+//
+// Implements cqlProtocolExtension interface.
```
Comment (mmatczuk) [inline scylla.go:56]: You can assert that by var _ cqlProtocolExtension = lwtAddMetadataMarkExt{}

**Comment by mmatczuk** [comment]
Thanks for detailed explanation, I'd just ask not to change `newFramer` for maintenance reasons, see comments inline otherwise it looks good.

File: scylla.go (hunk 5 above)
Comment (ManManson) [reply scylla.go:56]: Can you please explain why is this needed or where do I need to put that? If we use an object that does not implement all the methods from `cqlProtocolExtension` interface in places where it's expected, then the compiler will just issue a compiler error.

File: frame.go (hunk 1 above)
Comment (ManManson) [reply frame.go:409]: > It makes this patch unnecessarily big and you need to change all existing usages for no good reason.
> In other words this increases chances of conflicts with upstream changes.
> 
> I suggest to add a new constructor newFramerWithExts or use a plain setter that needs to be called before one starts using the framer.

Okay, I will adjust appropriately.

File: scylla.go [hunk 6]
```diff
@@ -4,17 +4,96 @@ import (
...
+
+// CQL Protocol extension interface for Scylla.
+// Each extension is identified by a name and defines a way to serialize itself
+// in STARTUP message payload.
+type cqlProtocolExtension interface {
+	name() string
+	serialize() map[string]string
+}
+
+func findCQLProtoExtByName(exts []cqlProtocolExtension, name string) *cqlProtocolExtension {
```
Comment (dahankzter) [inline scylla.go:31]: It's not very common to return a pointer to an interface in Go. That's the rationale for doing so here?

File: frame.go [hunk 7]
```diff
@@ -441,6 +443,18 @@ func newFramer(r io.Reader, w io.Writer, compressor Compressor, version byte) *f
 	return f
 }
 
+func newFramerWithExts(r io.Reader, w io.Writer, compressor Compressor, version byte,
+	cqlProtoExts []cqlProtocolExtension) *framer {
+
+	f := newFramer(r, w, compressor, version)
+
+	if lwtExt := findCQLProtoExtByName(cqlProtoExts, lwtAddMetadataMarkKey); lwtExt != nil {
+		f.flagLWT = (*lwtExt).(lwtAddMetadataMarkExt).lwtOptMetaBitMask
```
Comment (dahankzter) [inline frame.go:452]: This can blow up right? In case someone adds another thing implementing the interface this conversion blows up. 
Conversions in Go returns a complementary boolean that can be used to ensure panic free conversions.

Comment (ManManson) [reply frame.go:452]: In theory, it can, but is unlikely to happen since every implementation of the interface will have its own string constant, which identifies the extension. But I can add one more check for it, if you feel more comfortable with that.

File: scylla.go (hunk 6 above)
Comment (ManManson) [reply scylla.go:31]: What is the canonical way to lookup an object in a slice or map in Go? I can either return a pointer to the object or return an index, but in that case I would need to look it up again by doing `exts[i]` (probably doesn't incur any noticeable runtime cost, but anyway seems extraneous).

File: frame.go (hunk 7 above)
Comment (dahankzter) [reply frame.go]: I know these are in the category "I know the types being passed around" but this happened to me before in a long living project so I think handling it makes it both safer and more clear even if the probability of the error is very low.

File: scylla.go (hunk 6 above)
Comment (dahankzter) [reply scylla.go:31]: The returned pointer implements the interface right? Then you can declare the function to return the interface but the code returns the actual pointer to the concrete type.

Comment (ManManson) [reply scylla.go:31]: Thanks for the suggestion. I didn't realize that interfaces can have a somewhat mixed semantics, either a pointer or a value, so that you can assign a `nil` to a variable of interface type. I will adjust.

File: scylla.go [hunk 8]
```diff
@@ -4,17 +4,96 @@ import (
 	"fmt"
 	"math"
 	"strconv"
+	"strings"
```
Comment (dahankzter) [inline scylla.go:7]: I know these are frustrating but can you put all the formatting in a separate commit?

File: scylla.go (hunk 5 above)
Comment (dahankzter) [reply scylla.go:56]: It's a pretty common pattern to also declare and not just duck type the interface implementation.

File: scylla.go [hunk 9]
```diff
@@ -4,17 +4,96 @@ import (
...
+// increase LWT performance.
+//
+// Implements cqlProtocolExtension interface.
+type lwtAddMetadataMarkExt struct {
+	lwtOptMetaBitMask int
+}
+
+// Factory function to deserialize and create an `lwtAddMetadataMarkExt` instance
+// from SUPPORTED message payload.
+func newLwtAddMetaMarkExt(supported map[string][]string) *lwtAddMetadataMarkExt {
```
Comment (dahankzter) [inline scylla.go:63]: Why return a pointer here when you de reference it right away at the call site?

File: scylla.go [hunk 10]
```diff
@@ -71,6 +150,17 @@ func parseSupported(supported map[string][]string) scyllaSupported {
 	return si
 }
 
+func parseCQLProtocolExtensions(supported map[string][]string) []cqlProtocolExtension {
+	exts := []cqlProtocolExtension{}
+
+	lwtExt := newLwtAddMetaMarkExt(supported)
```
Comment (dahankzter) [inline scylla.go:156]: You can check if the name is empty if you return the struct instead of a pointer.

**Comment by dahankzter** [comment]
@mmatczuk any more comments?

File: scylla.go (hunk 5 above)
Comment (ManManson) [reply scylla.go:56]: OK, will do.

File: scylla.go (hunk 8 above)
Comment (ManManson) [reply scylla.go:7]: I don't think I follow. What exactly is not right here in terms of formatting?

File: scylla.go (hunk 9 above)
Comment (ManManson) [reply scylla.go:63]: What am I supposed to return from the function in case I cannot construct a proper `lwtAddMetadataMarkExt` object?

In this case I can return `nil` and employ a straightforward logic of checking for `nil` to assert that the object is ok.

In your case, AFAIU, it means that I would construct an `lwtAddMetadataMarkExt` with empty name, which I cannot do, if I am not to break the invariant of this type. 
Another approach is to either return an empty `lwtAddMetadataMarkExt{}` and check if the `lwtOptMetaBitMask` is empty (seems a little bit hackish) or return a pair (ext, error) and check for error instead (but it's not really an error if we could not find the extension key in `supported` map, so I don't think this would be good).

I am personally in favor of either the very first option or the last one

Comment (ManManson) [reply scylla.go:63]: But if it's something with "don't return pointers to structs/interfaces unless you really-really have to" rule, I can adjust.

File: scylla.go (hunk 8 above)
Comment (dahankzter) [reply scylla.go:7]: Nothing wrong other than it clutters the patch. I mean where you add a new dependency it's fine but I thought I saw an unrelated formatting. It's sometimes impossible to avoid if upstream also formats strangely. Ignore this if it's too messy or impossible to avoid.

Comment (ManManson) [reply scylla.go:7]: But I don't really see any unrelated formatting there, for me it looks aligned just as the rest of the surrounding lines...

Comment (ManManson) [reply scylla.go:7]: Maybe you can point me to a place where it's actually wrong?

File: conn.go [hunk 11]
```diff
@@ -423,6 +425,13 @@ func (s *startupCoordinator) startup(ctx context.Context) error {
 		}
 	}
 
+	for _, ext := range s.conn.cqlProtoExts {
+		var serialized = ext.serialize()
```
Comment (mmatczuk) [inline conn.go:429]: could use := syntax, var syntax it typically used when you declares variable without a value (nil pointer).

File: scylla.go (hunk 9 above)
Comment (mmatczuk) [reply scylla.go:64]: This type is very small

```
type lwtAddMetadataMarkExt struct {
	lwtOptMetaBitMask int
}
```

Plus in L90 it's `lwtAddMetadataMarkExt` that implements the interface not `*lwtAddMetadataMarkExt` so from this pov you could return `lwtAddMetadataMarkExt`.

Note that just the fact of having it accessed over interface would force compiler to escape that to heap.

My suggestion would be to change it so that `*lwtAddMetadataMarkExt` satisfies interface and leave this line as is.

**Comment by mmatczuk** [comment]
Overall looks good, see comments inline.

File: scylla.go (hunk 8 above)
Comment (mmatczuk) [reply scylla.go:7]: Run fmt on base branch and rebase your work on top.

Comment (ManManson) [reply scylla.go:7]: `gofmt` against the base branch shows empty output, so I assume everything was fine before my changes. I'll run another time after I've addressed all the issues and commit the result if there would be anything.

File: scylla.go (hunk 9 above)
Comment (ManManson) [reply scylla.go:64]: Hmm, it seems that I cannot duplicate implementing methods with both value and pointer receivers. So I assume it's easier to convert this function to return value intstead of a pointer, because otherwise it would require to change a lot more code.

Comment (mmatczuk) [reply scylla.go:64]: You cannot but compiler should do the work for you so there should be no code changes i.e. you can run function with a ptr receiver on struct.

Comment (ManManson) [reply scylla.go:64]: I think in that case I would need to adjust type assertion at the struct declaration to the following form:
```go
var _ cqlProtocolExtension = &lwtAddMetadataMarkExt{}
```
Which would effectively mean that I have a dangling object on the heap which cannot be recycled?
Or is the compiler smart enough to produce no code for this case?

Comment (dahankzter) [reply scylla.go:65]: I believe so since the `_` name makes unavailable for referencing.

Comment (ManManson) [reply scylla.go:65]: @dahankzter @mmatczuk Done.

**Comment by mmatczuk** [comment]
Thanks for the patch!
